### DISCUSSION TITLE
Question about 'index_select' in __lift__ function during graph propagation

### DISCUSSION THREAD

**Discussion by jaes77** [pyg-team/pytorch_geometric#4759] [Q&A]
Thanks for your efforts to develop such a great environment.

When launching GCN-like models with edge_index format, I noticed there is a step '__lift__' in __collect__ function during propagation.

https://github.com/pyg-team/pytorch_geometric/blob/09f25e9369b7f6f54e4a43413da697ddc6a77ac0/torch_geometric/nn/conv/message_passing.py#L217-L219

https://github.com/pyg-team/pytorch_geometric/blob/09f25e9369b7f6f54e4a43413da697ddc6a77ac0/torch_geometric/nn/conv/message_passing.py#L186-L189

It seems that the step is to arrange input tensor to different form with edge_index. But, I do not catch up what's going on during the step exactly because I'm not familiar with graph models.
What is src.index_select(node_dim, edge_index) exactly and why is the step needed only for 'torch.tensor' type?

I appreciate your support again.

**Comment by rusty1s** [comment]
The `lift` call will move node-level tensors to edge-level representations corresponding to source and destination nodes, e.g.
```python
x = ...  # [num_nodes, num_features]
x_src = x_j = x[edge_index[0]]  # [num_edges, num_features]
x_dst = x_i = x[edge_index[1]]  # [num_edges, num_features]
```
This is needed to compute a message per edge, which we then later aggregate according to destination node indices. Importantly, we can only lift `torch.Tensor` types since it is the only type that supports this kind of operation.
